index.js

```diff
@@ -55,6 +58,77 @@ app.get("/favicon.ico", (_req, res) => {
 app.all("/graphql", (req, res) => res.redirect("/"))
 
 app.use(bodyParser.json())
+
+function parse_args() {
+  return "( ... )"
+}
+
+function trace(res, span) {
+  span.addTags({
+    "http.status_code": res.statusCode,
+  })
+  span.finish()
+}
+
+app.use((req, res, next) => {
+  if (req.method === "POST") {
+    const tracer = new Tracer({ service: "metaphysics" })
+    const span = tracer.startSpan("metaphysics.query")
+    const query = req.body.query.replace(/(\()([^\)]*)(\))/g, parse_args)
+    span.addTags({
+      resource: query,
+      type: "web",
+      "span.kind": "server",
+      "http.method": req.method,
+      "http.url": req.url,
+    })
+
+    assign(req, { span })
+
+    res.on("finish", () => trace(res, span))
+    res.on("close", () => trace(res, span))
+  }
+  next()
+})
+
+function wrapResolve(typeName, fieldName, resolver) {
+  return function (root, options, request) {
+    const parentSpan = request.span
+    const span = parentSpan.tracer().startSpan("metaphysics.resolver." + typeName + "." + fieldName,
+      { childOf: parentSpan.context() })
+    span.addTags({
+      resource: typeName + ": " + fieldName,
+      type: "web",
+      "span.kind": "server",
+    })
+
+    assign(request, { span })
+    const result = resolver.apply(this, arguments);
+    assign(request, { span: parentSpan })
```

Comment:
Pass on what context? I also don't know much about graphqljs but this seemed like the only way I could pass some state to nested wrapped resolver call without modifying the resolver functions interface.

Reply:
Ah I see, good point @acjay. So in MP we’ve always incorrectly been using `rootValue` rather than `context`, but so that’s where you could store arbitrary data and your wrapper can take that out of there without needing to change any signatures.

Reply:
I'm also not fully fluent in GraphQL yet, but according to [the graphql-js docs](http://graphql.org/graphql-js/express-graphql/#graphqlhttp), I believe we can pass the spans via in an object passed along `context` parameter that we pass into `graphqlHTTP`.

I believe that would come through as the third argument of the resolver function, the one we're calling `request` here. Actually, I'm a bit confused, how is `request` currently being propogated at all? I was assuming it's the Express request object, but I don't see where we initially pass it in to the GraphQL executor, so I'm not sure.

Reply:
What I was trying to say is, please ignore `context` for now and use `rootValue` instead.

Reply:
Ah, here it is: https://github.com/graphql/express-graphql#options:

> context: A value to pass as the context to the graphql() function from GraphQL.js/src/execute.js. If context is not provided, the request object is passed as the context.

Do we rely on other aspects of `request` in our resolvers? If not, we could pass `{ span }` in for the `context` key in the argument to `graphqlHTTP`. I make this suggestion because it would avoid coupling everything to the request object, and we could avoid mutating it. It just seems to me like a safer approach. I'm a big fan of passing and mutating as little as possible, to the extent that that's practical.

Reply:
Again, please use `rootValue` for that for now. The change to use `context` properly is a larger work item.